astropy/utils/decorators.py

```diff
@@ -622,22 +622,16 @@ def __init__(self, fget, doc=None, lazy=False):
         if doc is not None:
             self.__doc__ = doc
 
-    def __get__(self, obj, objtype=None):
+    def __get__(self, obj, objtype):
         if self._lazy and objtype in self._cache:
             return self._cache[objtype]
 
-        if objtype is not None:
-            # The base property.__get__ will just return self here;
-            # instead we pass objtype through to the original wrapped
-            # function (which takes the class as its sole argument)
-            val = self.fget.__wrapped__(objtype)
-        else:
-            val = super().__get__(obj, objtype=objtype)
```

Comment:
It's not unheard of to call a descriptor's `__get__` directly.  I've had to do this from time to time in various cases (not necessarily with this one).  So if it ain't broke I wouldn't fix it.  But as you point out it *is* broke--but in this case I would just replace
```
val = super().__get__(obj, objtype=objtype)
```
with
```
val = super().__get__(obj, objtype)
```
and leave the rest alone.

Reply:
Ah, but it's a *bit* broken even in case one allows `objtype=None`. For example the first line checks `if self._lazy and objtype in self._cache:` but later inserts `objtype.__class__` in the `self._cache`. So this essentially bypasses `lazy`.

But setting `objtype = class(obj)` if `objtype is None` essentially also makes the removed `if`s here obsolete.

It may be of limited use when one uses `classproperty` as Mixin, when `super()` doesn't resolve to `property` (but then why would normal access which always sets `objtype` to a not-None value **not** respect inheritance but direct access with `__get__` and not giving the second argument or setting it to `None` would).

Maybe you're right and one shouldn't fix what ain't broken. But I accidentally stumbled over it when reviewing another PR and it currently seems over-engineered and broken in several places (that won't be accessible by normal means).

Reply:
I totally forgot: Thanks for the comment. I'm very glad you had the time 👍